app/src/main/java/com/ddoczi/tasky/authentication/presentation/login/LoginViewModel.kt

```diff
@@ -1,45 +1,46 @@
 package com.ddoczi.tasky.authentication.presentation.login
 
-import androidx.compose.runtime.mutableStateOf
-import androidx.compose.runtime.getValue
-import androidx.compose.runtime.setValue
 import androidx.lifecycle.ViewModel
-import com.ddoczi.tasky.authentication.domain.EmailValidator
-import com.ddoczi.tasky.authentication.domain.PasswordValidator
+import com.ddoczi.tasky.authentication.domain.AuthDataValidator
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import javax.inject.Inject
 
-// TODO add DI and update validation
-class LoginViewModel(): ViewModel() {
-    var state by mutableStateOf(LoginState())
-        private set
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val authDataValidator: AuthDataValidator
+): ViewModel() {
+    private val _state = MutableStateFlow(LoginState())
+    val state = _state.asStateFlow()
 
     fun onEvent(event: LoginEvent) {
         when (event) {
             is LoginEvent.OnEmailChanged -> {
-                state = state.copy(
+                _state.value = _state.value.copy(
                     email = event.email,
-                    isEmailValid = EmailValidator().isEmailValid(event.email),
-                    emailError = false
+                    isEmailValid = authDataValidator.validEmail(event.email),
+                    isEmailError = false
                 )
             }
             is LoginEvent.OnPasswordChanged -> {
-                state = state.copy(
+                _state.value = _state.value.copy(
                     password = event.password,
-                    isPasswordValid = PasswordValidator().isPasswordValid(event.password),
-                    passwordError = false
+                    isPasswordError = false
                 )
             }
             is LoginEvent.OnPasswordVisibilityToggle -> {
-                state = state.copy(isPasswordVisible = !state.isPasswordVisible)
+                _state.value = _state.value.copy(isPasswordVisible = !_state.value.isPasswordVisible)
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue, but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.
Using `_state.value = _state.value.copy(...)` is fine if you always call this from a single thread (which you do here). But in ViewModels, you're very likely to not do that as soon as you'd change the state inside a `viewModelScope`.